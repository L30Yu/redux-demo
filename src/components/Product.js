import React from "react";
import { Input, Button } from "reactstrap";

const style = {
  padding: "1em 1em",
};

export default class Product extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      quantity: 1,
    };
    this.onChange = this.onChange.bind(this);
  }

  onChange = event => {
    let quantity = parseInt(event.target.value, 10);
    quantity = Number.isNaN(quantity) || quantity < 0 ? 0 : quantity;
    if (quantity > this.props.product.inventory) {
      quantity = this.props.product.inventory;
    }
    this.setState({
      quantity
    });
  };

  onAddClick = (id, quantity) => {
    if (quantity > this.props.product.inventory - this.props.orderCount) {
      quantity = this.props.product.inventory - this.props.orderCount;
    }
    if (quantity <= 0) return;
    this.props.addToBasket(id, quantity);
  }

  onRemoveClick = (id, quantity) => {
    if (quantity > this.props.orderCount) {
      quantity = this.props.orderCount;
    }
    if (quantity <= 0) return;
    this.props.removeFromBasket(id, quantity);
  }

  render() {
    let { product, orderCount } = this.props;
    let disableRemoveBtn = orderCount == null || orderCount <= 0;
    let disableAddBtn = product.inventory <= 0 || product.inventory <= orderCount;
    return (
        <tr style={style}>
          <td>
            {product.name}
          </td>
          <td>
            <Input invalide={this.state.invalide} type="number" name="quantity" placeholder="quantity" id={product.id} value={this.state.quantity} onChange={this.onChange} />
          </td>
          <td>
            <Button color="primary" onClick={() => this.onAddClick(product.id, this.state.quantity)} disabled={disableAddBtn}>
              Add
          </Button>
          </td>
          <td>
            <Button color="danger" onClick={() => this.onRemoveClick(product.id, this.state.quantity)} disabled={disableRemoveBtn}>
              Remove
          </Button>
          </td>
        </tr>
    );
  }
}