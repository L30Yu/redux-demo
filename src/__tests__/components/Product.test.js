import React from 'react';
import Product from '../../components/Product';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
    let orderCount = 2;
    let product = {
        id: 0,
        name: 'test',
        price: 2
    }
    const tree = renderer
        .create(<Product product={product} orderCount={orderCount} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
