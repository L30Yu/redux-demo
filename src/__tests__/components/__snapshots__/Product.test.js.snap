// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<tr
  style={
    Object {
      "padding": "1em 1em",
    }
  }
>
  <td>
    test
  </td>
  <td>
    <input
      className="form-control"
      id={0}
      invalide={undefined}
      name="quantity"
      onChange={[Function]}
      placeholder="quantity"
      type="number"
      value={1}
    />
  </td>
  <td>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </td>
  <td>
    <button
      className="btn btn-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </td>
</tr>
`;
